### PR TITLE
Add custom dataset training with labels and embeddings visualization example

### PR DESCRIPTION
**Pull request by michaelmendelson**
Added script to train on custom dataset with frame level or sequence level labels. Modified dataset class so that you feed in annotations. During eval the script will flexibly use either classification or regression for either type of label based on the 'eval_utils' metadata that you also enter into the dataset. Showed PCA visualization of latents in the jupyter notebook.

### CODE REVIEW

File: bams/__init__.py [hunk 1]
```diff
@@ -1,8 +1,9 @@
 from .hoa_loss import HoALoss
+from .load_latents import compute_representations
 # from . import human_bandit
 # from . import legged_robots
 # from . import mabe
 # from .dataset import Dataset
 # from . import models
```
Comment (mazabou) [inline bams/__init__.py]: rm

File: README.md [hunk 2]
```diff
@@ -65,8 +65,8 @@ cd mabe-2022-public-evaluator/
 python3 round1_evaluator.py --task mouse --submission ../bams-mouse-triplet-2023-11-30-17-49-59_submission.npy --labels ../data/mabe/mouse_triplets_test_labels.npy
 ```
 
-## Custom datasets
-If you'd like to test BAMS on your own dataset, please use `custom_dataset.py`. 
+## Custom datasets without annotations
+If you'd like to train BAMS on your own dataset without behavioral readouts, please use `custom_dataset.py`.
```
Comment (mazabou) [inline README.md]: rename

File: bams/data/dataset.py [hunk 3]
```diff
@@ -39,6 +39,8 @@ def __init__(
         hoa_window=30,
         cache_path=None,
         cache=False,
+        annotations=None,
```
Comment (mazabou) [inline bams/data/dataset.py:46]: add documentation

**Review by mazabou — APPROVED** [review]
Looks good, thank you!

File: custom_dataset_w_labels.py [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+            device,
+            train_loader,
+            optimizer,
+            criterion,
+            writer,
+            step,
+            args.log_every_step,
+        )
+        scheduler.step()
+        torch.save(model.state_dict(), model_name + ".pt")
```
Comment (michaelmendelson) [inline custom_dataset_w_labels.py]: remove (repetitive) move subsequent save BEFORE test()

File: custom_dataset_w_labels.py [hunk 5]
```diff
@@ -0,0 +1,259 @@
...
+
+        if global_pool:
+            emb = torch.mean(emb, dim=1, keepdim=True)
+
+        train_data = [emb[train_idx].reshape(-1, emb_size), target[train_idx].reshape(-1, 1)]
+        test_data = [emb[test_idx].reshape(-1, emb_size), target[test_idx].reshape(-1, 1)]
+        mse = train_linear_regressor(train_data, test_data, device, lr=1e-2, weight_decay=1e-4)
+        return mse
+    
+    data = next(iter(DataLoader(dataset, batch_size=len(dataset), shuffle=False)))
```
Comment (michaelmendelson) [inline custom_dataset_w_labels.py]: compute inefficient change to 
```suggestion
    keys = list(dataset.annotations.keys())
    data = {}
    for batch in DataLoader(dataset, batch_size=4, shuffle=False):
      for key in keys:
        if len(data.keys()) < len(keys):
          data[key] = batch[key]
        else:
          data[key] = torch.cat([data[key], batch[key]], axis=0)
```

File: custom_dataset_w_labels.py [hunk 6]
```diff
@@ -0,0 +1,259 @@
...
+    for target_tag in dataset.eval_utils['classification_tags']:
+        target = torch.LongTensor(data[target_tag])
```
Comment (michaelmendelson) [inline custom_dataset_w_labels.py]:
```suggestion
        target = data[target_tag].type(torch.LongTensor)
```
